File: backend/app/services/storage.py
import uuid
from typing import BinaryIO

from app.db import get_supabase

BUCKET = "report-images"


def upload_report_image(file: BinaryIO, filename: str, content_type: str) -> str:
    ext = filename.rsplit(".", 1)[-1] if "." in filename else "jpg"
    object_name = f"report_{uuid.uuid4().hex}.{ext}"

    supabase = get_supabase()
    file_bytes = file.read()

    supabase.storage.from_(BUCKET).upload(
        object_name,
        file_bytes,
        file_options={"content-type": content_type, "cache-control": "3600", "upsert": "false"},
    )

    public_url = supabase.storage.from_(BUCKET).get_public_url(object_name)
    return public_url

File: backend/src/services/storage.js
import crypto from 'crypto'
import { getSupabase } from '../db.js'

const BUCKET = 'report-images'

/**
 * Upload a report image to Supabase Storage bucket and return public URL
 * @param {Buffer} fileBuffer
 * @param {string} originalFilename
 * @param {string} contentType
 * @returns {Promise<string>}
 */
export async function uploadReportImage(fileBuffer, originalFilename, contentType = 'image/jpeg') {
  const ext = originalFilename.includes('.') ? originalFilename.split('.').pop() : 'jpg'
  const objectName = `report_${crypto.randomUUID().replace(/-/g, '')}.${ext}`

  const supabase = getSupabase()

  const { error: uploadError } = await supabase.storage.from(BUCKET).upload(objectName, fileBuffer, {
    contentType: contentType || 'image/jpeg',
    cacheControl: '3600',
    upsert: false,
  })

  if (uploadError) {
    throw new Error(`Failed to upload image to storage: ${uploadError.message}`)
  }

  const { data } = supabase.storage.from(BUCKET).getPublicUrl(objectName)
  return data.publicUrl
}
